### PR TITLE
 [GSB] Avoid emitting associated type diagnostics on the protocol decl

### PR DESCRIPTION
**Pull request by hamishknight**
For explicit abstract protocol floating requirement sources, get the source location from the protocol requirement rather than delegating to the parent `RequirementSource`. This allows us to diagnose on an associated type requirement directly rather than on the protocol declaration.

(Also, clean up a few unnecessary `TypeLoc::withoutLoc()`s)

### CODE REVIEW

**Comment by jrose-apple** [comment]
Great idea, but can't RequirementSource provide this information itself? There's a note in RequirementSource::getLoc about why it's not doing it, but I don't understand it myself.

File: test/NameBinding/scope_map_lookup.swift [hunk 1]
```diff
@@ -23,9 +23,10 @@ protocol P1 {
 }
 
 // Protocols involving associated types.
-protocol AProtocol { // expected-error{{type 'Self.e' constrained to non-protocol, non-class type 'Self.e'}}
+protocol AProtocol {
   associatedtype e : e
   // expected-error@-1 {{inheritance from non-protocol, non-class type 'Self.e'}}
+  // expected-error@-2 {{type 'Self.e' constrained to non-protocol, non-class type 'Self.e'}}
```
Comment (DougGregor) [inline test/NameBinding/scope_map_lookup.swift]: Much better!

**Review by DougGregor — APPROVED** [review]
Nice improvement, thanks

File: lib/AST/GenericSignatureBuilder.cpp [hunk 2]
```diff
@@ -1678,22 +1678,34 @@ const RequirementSource *FloatingRequirementSource::getSource(
 }
 
 SourceLoc FloatingRequirementSource::getLoc() const {
+  auto getSeparatorLoc = [](const RequirementRepr *req) -> SourceLoc {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/AST/GenericSignatureBuilder.cpp]: Good idea! In that case, I think we might as well remove `getColonLoc()` and `getEqualLoc()` in favour of `getSeparatorLoc()`, as most callers aren't concerned about the difference (and those that are are already exercising the appropriate assertion with e.g the use of `getFirstType()`).

[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/AST/GenericSignatureBuilder.cpp]: Ping @DougGregor

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by hamishknight** [comment]
Is this good to go?

**Pull request merged**